Donation-platform/src/test/java/com/google/PersonalizedRecommendationsTest.java

```diff
@@ -11,269 +11,340 @@
     int indexOfLowestRankingCharityWithTag1 = 0;
     for (Charity charity : actual) {
-      if ((new HashSet<>(charity.getTags())).contains("children")
+      Collection<Tag> charityTagObjects = charity.getCategories();
+      Collection<String> charityTagNames = new ArrayList<String>();
+      for(Tag tag : charityTagObjects) {
+        charityTagNames.add(tag.getName());
+      }
+      if ((new HashSet<>(charityTagNames)).contains("children")
           && actual.indexOf(charity) > indexOfLowestRankingCharityWithTag1) {
         indexOfLowestRankingCharityWithTag1 = actual.indexOf(charity);
       }
     }
     correctOrdering = indexOfLowestRankingCharityWithTag1 < indexOfHighestRankingCharityWithoutTag1;
-
+ 
     Assert.assertTrue(everyCharityPresent && correctOrdering);
   }
-
+ 
   @Test
   public void onlyTags2and3Selected() {
     // If only tags #2 and #3 are selected, recommend all charities that have tag
     // #3, tag #3, or both #2 and #3. Every charity that has tag #2 should come before
     // every charity that does not have tag #2.
-
+ 
     // Example: All charities with the tags "health" and/or "children" should be
     // recommended. Every charity with the tag "health" should come before every
     // charity without the tag "health."
-
+ 
     boolean everyCharityPresent = false;
     boolean correctOrdering = false;
-
+ 
     List<String> tags = Arrays.asList(null, "health", "children");
     List<Charity> actual = recommendation.getBestMatches(tags);
     List<Charity> expected = Arrays.asList(SJ, RC, AHA, YMCA);
-    everyCharityPresent = (new HashSet<>(expected)).equals(new HashSet<>(actual));
-
+    everyCharityPresent = sameSetOfCharities(actual, expected);
+ 
     // Find the index of the highest-ranking charity without tag #2
     int indexOfHighestRankingCharityWithoutTag2 = actual.size() - 1;
     for (Charity charity : actual) {
-      if (!(new HashSet<>(charity.getTags())).contains("health")
+      Collection<Tag> charityTagObjects = charity.getCategories();
+      Collection<String> charityTagNames = new ArrayList<String>();
+      for(Tag tag : charityTagObjects) {
+        charityTagNames.add(tag.getName());
+      }
+      if (!(new HashSet<>(charityTagNames)).contains("health")
           && actual.indexOf(charity) < indexOfHighestRankingCharityWithoutTag2) {
         indexOfHighestRankingCharityWithoutTag2 = actual.indexOf(charity);
       }
     }
     // Find the index of the lowest-ranking charity with tag #2
     int indexOfLowestRankingCharityWithTag2 = 0;
     for (Charity charity : actual) {
-      if ((new HashSet<>(charity.getTags())).contains("health")
+      Collection<Tag> charityTagObjects = charity.getCategories();
+      Collection<String> charityTagNames = new ArrayList<String>();
+      for(Tag tag : charityTagObjects) {
+        charityTagNames.add(tag.getName());
+      }
+      if ((new HashSet<>(charityTagNames)).contains("health")
```

Comment:
do we need to cast it here to hashset? Because i see you removing above at some places. Do we need to remove duplicates before checking and so you are casting it?

Reply:
I am not removing any duplicates as there are none. I forgot that Collections has a contains method; that's why I converted charityTagNames into a HashSet. I have removed these unnecessary casts.